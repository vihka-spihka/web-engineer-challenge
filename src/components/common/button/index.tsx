import React from 'react';

interface IButton {
	label: string;
	onClick: (e: React.SyntheticEvent) => void;
};

export const Button = ({ label, onClick }: IButton) => {
	return (
		<button onClick={() => onClick}>
			{label}
		</button>
	);
};
