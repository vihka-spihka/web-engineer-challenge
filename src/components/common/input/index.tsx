import React from 'react';
import { StyledInput } from './../styles';

interface IInput extends React.InputHTMLAttributes<HTMLInputElement> {
	label: string;
	onChangeInput: (v: InputValue) => void;
};

export const Input: React.FC<IInput> = ({ label, value, onChangeInput, ...props }) => {
	return (
		<StyledInput>
			<label>
				{label}
				<input
					value={value}
					onChange={
						(e: React.ChangeEvent<HTMLInputElement>) => onChangeInput(e.target.value)
					}
					{...props} />
			</label>
		</StyledInput>
	);
};
