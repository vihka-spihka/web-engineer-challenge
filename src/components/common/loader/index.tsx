import React from 'react';
import { StyledLoader } from './../styles';

export const Loader: React.FC = () => {
	return (
		<StyledLoader>
			<img src='/favicon.ico' width='80' height='80' alt='Company logo' />
		</StyledLoader>
	);
};
