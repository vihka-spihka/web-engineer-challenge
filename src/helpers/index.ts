const fetchData = (
	url: string,
	method: string,
	headers?: {[key: string]: string}
) => {
	return fetch(url, {
		method,
		headers,
	}).then(res => res.json());
};

export default { fetchData };
