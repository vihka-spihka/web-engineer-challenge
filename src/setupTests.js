// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

export const defaultTheme = {
	data: {
		global: {
			body_color: 'green',
			loading_bar_color: 'red',
			background_color: 'white',
			primary_border_color: 'black'
		},
		primary_font_family: {
			regular: 'Font',
			bold: 'Font',
		},
		buttons: {
			primary_button_text_color: 'blue',
			primary_button_font_family: 'blue',
			primary_button_border_color: 'blue',
			primary_button_border_radius: 10,
			primary_button_background_color: 'blue',
			secondary_button_text_color: 'blue',
			secondary_button_font_family: 'blue',
			secondary_button_border_color: 'blue',
			secondary_button_background_color: 'blue',
			secondary_button_border_radius: 10,
			knockout_button_text_color: 'blue',
			knockout_button_font_family: 'blue',
			knockout_button_border_color: 'blue',
			knockout_button_background_color: 'blue',
			knockout_button_border_radius: 10,
		},
		forms: {
			field_height: 50,
			placeholder_color: 'black',
		},
		header: {
			font_color: 'blue',
			border_color: 'red',
			border_radius: 9,
			background_color: 'green',
			desktop_logo_width: 200,
			font_size: 10,
		}
	},
};

Enzyme.configure({ adapter: new Adapter() });
