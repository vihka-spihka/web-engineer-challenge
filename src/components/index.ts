import { Input, StyledButton, Loader } from './common';
import { Countdown } from './countdown';
import { Footer } from './footer';
import { Header } from './header';
import { History } from './history';

export { Input, StyledButton, Loader, Countdown, Footer, Header, History };
