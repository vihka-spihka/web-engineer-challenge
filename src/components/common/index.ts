import { Input } from './input';
import { StyledButton } from './styles';
import { Loader } from './loader';

export { Input, StyledButton , Loader};
