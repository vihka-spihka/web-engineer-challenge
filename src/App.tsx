import React from 'react';
import { ThemeProvider } from 'styled-components';
import { StyledApp } from './styles';
import { Countdown } from './components/countdown';
import { Loader } from './components/common';
import { Header } from './components/header';
import { Footer } from './components/footer';
import helpers from './helpers';

export enum COUNTDOWN_MODE {
	NEW,
	EDIT,
	IN_PROGRESS,
};

interface IState {
	mode: COUNTDOWN_MODE
  theme: Theme | null;
  countdown: Countdown | null;
}

/* TODO: think about the implementation of IProps as { defaultTheme: ITheme } */
class App extends React.Component<unknown, IState> {
  state: IState = {
  	mode: COUNTDOWN_MODE.NEW,
  	theme: null,
  	countdown: null,
  }

  componentDidMount(): void {
  	helpers.fetchData(
  		'https://api.koala.io/marketing/v1/device-configurations/alias/web-config',
  		'GET',
  		{'X-Organization-Id': '1'}
  	)
  		.then(result => this.setState({ theme: result.data }))
  		.catch(error => console.log(error.message));
  }

  render() {
  	const data = this.state.theme && this.state.theme.data;
  	const { theme, mode, countdown } = this.state;
  	return (
  		!theme
  			? <Loader />
  			: <ThemeProvider theme={data}>
  				<StyledApp />
  				<Header />
  				<main>
  					{
  						!countdown && mode === COUNTDOWN_MODE.IN_PROGRESS
  							? <Loader />
  							: <Countdown
  								mode={mode}
  								timeLeft={countdown?.timeLeft}
  								changeMode={this.changeMode} />
  					}
  				</main>
  				<Footer />
  			</ThemeProvider>
  	);
  }

  setTimer = (time: InputValue): void => {
  	const countdown = setInterval(() => {
  		!this.state.countdown
  			? this.setState({
  				countdown: {
  					id: countdown,
  					start: Date.now(),
  					duration: +time,
  					timeLeft: +time - 1,
  				}
  			})
  			: this.setState({
  				countdown: {
  					...this.state.countdown,
  					timeLeft: this.state.countdown.timeLeft - 1
  				}});

  		if (this.state.countdown?.timeLeft === 0) {
  			this.changeMode(COUNTDOWN_MODE.NEW);
  		}
  	}, 1000);
  };

	changeMode = (mode: COUNTDOWN_MODE, time?: number): void => {
		this.setState({mode});
		switch (mode) {
			case COUNTDOWN_MODE.IN_PROGRESS:
				this.setTimer(time as number);
				break;
			case COUNTDOWN_MODE.EDIT:
				const timer_id = this.state.countdown?.id;
				clearInterval(timer_id);
				this.setState({countdown: null});
				break;
			case COUNTDOWN_MODE.NEW:
			default:
				clearInterval(this.state.countdown?.id);
				this.setState({countdown: null});
				break;
		};
	};
}

export default App;
