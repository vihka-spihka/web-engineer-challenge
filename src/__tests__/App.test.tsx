import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import App, { COUNTDOWN_MODE } from '../App';
import { Loader } from './../components/common';
import helpers from './../helpers';
import { defaultTheme } from './../setupTests';

const countdown = {
	id: 1,
	start: Date.now(),
	duration: 10,
	timeLeft: 10
};

let appWrapper: ShallowWrapper<any, any, App>;
let instance: App;

const mockSuccessfulFetch = (): Promise<any> => {
	return Promise.resolve({data: defaultTheme});
};

const mockFailedFetch = (): Promise<any> => {
	return Promise.reject({message: 'Error'});
};

describe('App component', () => {
	beforeEach(() => {
		appWrapper = shallow<App>(<App />, {disableLifecycleMethods: true});
		instance = appWrapper.instance();
		appWrapper.setState({theme: null, countdown: null});

		helpers.fetchData = jest.fn().mockImplementation(mockSuccessfulFetch);
	});

	it('renders Loader when no theme provided', () => {
		expect(appWrapper.contains(<Loader />)).toEqual(true);
		expect(appWrapper).toMatchSnapshot();
	});

	it('renders correctly when theme provided', () => {
		instance.setState({theme: defaultTheme});
		expect(appWrapper).not.toEqual('');
		expect(appWrapper).toMatchSnapshot();
	});

	it('renders correctly when countdown is in progress', () => {
		instance.setState({countdown: countdown, theme: defaultTheme});
		expect(appWrapper.contains(<Loader />)).toEqual(false);
		expect(appWrapper).toMatchSnapshot();
	});

	it('shows Loader when countdown is about to start', () => {
		instance.setState({
			countdown: null,
			theme: defaultTheme,
			mode: COUNTDOWN_MODE.IN_PROGRESS
		});
		expect(appWrapper.contains(<Loader />)).toEqual(true);
	});

	it('shows `Seconds left:` if countdown is in progress', () => {
		instance.setState({
			countdown: countdown,
			theme: defaultTheme,
			mode: COUNTDOWN_MODE.IN_PROGRESS
		});
		expect(appWrapper.find('Countdown').length).toEqual(1);
	});

	it('sets state.theme on successful fetch in componentDidMount', () => {
		instance.componentDidMount!();
		// waiting for the state to update
		return new Promise(res => setImmediate(res))
			.then(() => {
				expect(appWrapper.state()).toMatchObject({theme: defaultTheme});
			});
	});

	it('doesn`t modify state.theme on failed fetch in componentDidMount', () => {
		helpers.fetchData = jest.fn().mockImplementation(mockFailedFetch);
		instance.componentDidMount!();
		// waiting for the state to update
		return new Promise(res => setImmediate(res))
			.then(() => {
				expect(appWrapper.state()).toMatchObject({theme: null});
			});
	});

	it('sets countdown for 10 seconds', () => {
		jest.useFakeTimers();
		instance.setTimer(10);
		// use 1000ms due to the state hasn't been refreshed immediatelly
		jest.advanceTimersByTime(1000);

		expect(appWrapper.state()).toMatchObject({
			countdown: {
				duration: 10
			}
		});
	});

	it('sets mode correctly', () => {
		const modes = [COUNTDOWN_MODE.NEW, COUNTDOWN_MODE.IN_PROGRESS, COUNTDOWN_MODE.EDIT];
		modes.forEach((it:COUNTDOWN_MODE) => {
			instance.changeMode(it);
			expect(appWrapper.state()).toMatchObject({ mode: it });
		});
	});

	it('removes coundown when time is up', () => {
		jest.useFakeTimers();
		instance.setTimer(10);
		jest.advanceTimersByTime(11000);

		expect(appWrapper.state()).toMatchObject({countdown: null});
	});

	afterAll(() => {
		appWrapper.unmount();
		jest.unmock('./../helpers');
		jest.clearAllTimers();
	});
});
