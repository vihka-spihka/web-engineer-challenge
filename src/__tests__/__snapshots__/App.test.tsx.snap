// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders Loader when no theme provided 1`] = `<Loader />`;

exports[`App component renders correctly when countdown is in progress 1`] = `
<Component
  theme={
    Object {
      "buttons": Object {
        "knockout_button_background_color": "blue",
        "knockout_button_border_color": "blue",
        "knockout_button_border_radius": 10,
        "knockout_button_font_family": "blue",
        "knockout_button_text_color": "blue",
        "primary_button_background_color": "blue",
        "primary_button_border_color": "blue",
        "primary_button_border_radius": 10,
        "primary_button_font_family": "blue",
        "primary_button_text_color": "blue",
        "secondary_button_background_color": "blue",
        "secondary_button_border_color": "blue",
        "secondary_button_border_radius": 10,
        "secondary_button_font_family": "blue",
        "secondary_button_text_color": "blue",
      },
      "forms": Object {
        "field_height": 50,
        "placeholder_color": "black",
      },
      "global": Object {
        "background_color": "white",
        "body_color": "green",
        "loading_bar_color": "red",
        "primary_border_color": "black",
      },
      "header": Object {
        "background_color": "green",
        "border_color": "red",
        "border_radius": 9,
        "desktop_logo_width": 200,
        "font_color": "blue",
        "font_size": 10,
      },
      "primary_font_family": Object {
        "bold": "Font",
        "regular": "Font",
      },
    }
  }
>
  <Memo(l) />
  <Header />
  <main>
    <Countdown
      changeMode={[Function]}
      mode={0}
      timeLeft={10}
    />
  </main>
  <Footer />
</Component>
`;

exports[`App component renders correctly when theme provided 1`] = `
<Component
  theme={
    Object {
      "buttons": Object {
        "knockout_button_background_color": "blue",
        "knockout_button_border_color": "blue",
        "knockout_button_border_radius": 10,
        "knockout_button_font_family": "blue",
        "knockout_button_text_color": "blue",
        "primary_button_background_color": "blue",
        "primary_button_border_color": "blue",
        "primary_button_border_radius": 10,
        "primary_button_font_family": "blue",
        "primary_button_text_color": "blue",
        "secondary_button_background_color": "blue",
        "secondary_button_border_color": "blue",
        "secondary_button_border_radius": 10,
        "secondary_button_font_family": "blue",
        "secondary_button_text_color": "blue",
      },
      "forms": Object {
        "field_height": 50,
        "placeholder_color": "black",
      },
      "global": Object {
        "background_color": "white",
        "body_color": "green",
        "loading_bar_color": "red",
        "primary_border_color": "black",
      },
      "header": Object {
        "background_color": "green",
        "border_color": "red",
        "border_radius": 9,
        "desktop_logo_width": 200,
        "font_color": "blue",
        "font_size": 10,
      },
      "primary_font_family": Object {
        "bold": "Font",
        "regular": "Font",
      },
    }
  }
>
  <Memo(l) />
  <Header />
  <main>
    <Countdown
      changeMode={[Function]}
      mode={0}
    />
  </main>
  <Footer />
</Component>
`;
